### PR TITLE
Describe how to convert non-string primitives for protocols which only support strings

### PR DESCRIPTION
**Pull request by dashpole**
## Changes

Small change forked from https://github.com/open-telemetry/opentelemetry-specification/pull/2266#discussion_r804900011.  Similarly to specifying how to represent arrays, we should specify how to represent primitives as strings if those are not supported.  We should use the same string representation for primitive types as we do for primitives within arrays.

I'm not sure if this is actually allowed, since the spec is stable.  But since this is a SHOULD, rather than a MUST, it wouldn't require breaking changes.  If it isn't permitted, I can just document this in the Prometheus conversion spec itself.

@bogdandrutu I think this would require slight changes to https://github.com/open-telemetry/opentelemetry-collector/blob/30e6bf6eeee544e869411bf50c57447323d259d9/model/pdata/common.go#L370.

cc @jmacd @jsuereth

### CODE REVIEW

File: specification/common/common.md [hunk 1]
```diff
@@ -24,10 +24,10 @@ An `Attribute` is a key-value pair, which MUST have the following properties:
 
 - The attribute key MUST be a non-`null` and non-empty string.
 - The attribute value is either:
-  - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer.
+  - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer. For protocols that do not natively support non-string values, such values SHOULD be represented as JSON strings.
```
[the remark this answers is not in the document]
Comment (dashpole) [reply specification/common/common.md:27]: Ah, yeah my wording isn't clear.  "such values" is mean to reference the not "support[ed] non-string values" from the first part of the sentence.  I'll update it.

My goal was unambiguous representation, so that (e.g. prometheus) exporters use the same value for identical labels across language SDKs + collector.

[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply specification/common/common.md]: >Ah, yeah my wording isn't clear. "such values" is mean to reference the not "support[ed] non-string values" from the first part of the sentence. I'll update it.

SGTM

>My goal was unambiguous representation

The wording in this PR results in an ambiguous representation though, right? What does `true` mean in the resulting non-OTLP format? Is it a boolean `true` or a string `true` which is missing the surrounding quotes?

To be clear: I don't think this is a problem. We likely don't care about this sort of ambiguity since we do not expect the reverse translation to be applied (from non-OTLP format to OTLP) and to be expected to yield the original data with correct data types.

TLDR; your change looks good to me.

[the remark this answers is not in the document]
Comment (dashpole) [reply specification/common/common.md]: Ah, I get your point.  I want it to be unambiguous how int64(100) is represented as a string, but I don't need to be able to tell if "100" was initially a string or an int.

File: specification/common/common.md [hunk 2]
```diff
@@ -24,10 +24,10 @@ An `Attribute` is a key-value pair, which MUST have the following properties:
 
 - The attribute key MUST be a non-`null` and non-empty string.
 - The attribute value is either:
-  - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer.
+  - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer. For protocols that do not natively support non-string values, non-string values SHOULD be represented as JSON strings.
   - An array of primitive type values. The array MUST be homogeneous,
     i.e., it MUST NOT contain values of different types. For protocols that do
-    not natively support array values such values SHOULD be represented as JSON strings.
+    not natively support array values, such values SHOULD be represented as JSON strings.
```
Comment (tigrannajaryan) [inline specification/common/common.md:30]: Maybe eliminate the duplication and only say it once after these 2 bullet points:
```
For protocols that do not natively support non-string values, non-string values SHOULD be represented as JSON strings.
```
This covers the arrays as well (it is a non-string value).

**Review by tigrannajaryan — APPROVED** [review]

File: specification/common/common.md (hunk 2 above)
Comment (dashpole) [reply specification/common/common.md]: Done.

**Review by tigrannajaryan — APPROVED** [review]

File: specification/common/common.md [hunk 3]
```diff
@@ -26,8 +26,9 @@ An `Attribute` is a key-value pair, which MUST have the following properties:
 - The attribute value is either:
   - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer.
   - An array of primitive type values. The array MUST be homogeneous,
-    i.e., it MUST NOT contain values of different types. For protocols that do
-    not natively support array values such values SHOULD be represented as JSON strings.
+    i.e., it MUST NOT contain values of different types.
+
+For protocols that do not natively support non-string values, non-string values SHOULD be represented as JSON strings.
```
Comment (jmacd) [inline specification/common/common.md:31]:
```suggestion
For protocols that do not natively support non-string values, non-string values SHOULD be represented as JSON-encoded strings.  For example, the expression `int64(100)` will be encoded as `100`, `float64(1.5)` will be encoded as `1.5`, and an empty array of any type will be encoded as `[]`.
```

**Review by jmacd — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
FYI, I have a related open OTEP which tells how to convert from values to OTLP value (the opposite direction of what this PR defines): https://github.com/open-telemetry/oteps/pull/178

File: specification/common/common.md [hunk 4]
```diff
@@ -26,8 +26,9 @@ An `Attribute` is a key-value pair, which MUST have the following properties:
 - The attribute value is either:
   - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer.
   - An array of primitive type values. The array MUST be homogeneous,
-    i.e., it MUST NOT contain values of different types. For protocols that do
-    not natively support array values such values SHOULD be represented as JSON strings.
+    i.e., it MUST NOT contain values of different types.
+
+For protocols that do not natively support non-string values, non-string values SHOULD be represented as JSON-encoded strings.  For example, the expression `int64(100)` will be encoded as `100`, `float64(1.5)` will be encoded as `1.5`, and an empty array of any type will be encoded as `[]`.
```
[the remark this answers is not in the document]
Comment (dashpole) [reply specification/common/common.md:31]: Are attributes allowed to be bytes?  It isn't currently listed above.  I would encode it as `bb` if we supported them.

I think we SHOULD sort keys in ascending order, since JSON doesn't require encoders to order keys.  But it looks like most languages do support sorting.

[the remark this answers is not in the document]
Comment (dashpole) [reply specification/common/common.md:31]: Wait, I don't see maps as a permitted attribute type either.

[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply specification/common/common.md:31]: This PR is for implementing the exporters, right? In that case bytes or maps are irrelevant since they are not allowed by the API.

[the remark this answers is not in the document]
Comment (aabmass) [reply specification/common/common.md:31]: Would this not apply to collector exporters converting to OTLP? The file this is updating doesn't say it's specific to SDK exporters, am I missing something?

[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply specification/common/common.md:31]: @aabmass that's a good point. I incorrectly assumed the updated document was in the SDK section, but it is not, it is in the common which applies to all of OpenTelemetry.

[the remark this answers is not in the document]
Comment (dashpole) [reply specification/common/common.md:31]: Just to confirm, I should update this spec to include bytes and maps as valid attribute values?

[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply specification/common/common.md:31]: Perhaps in a separate PR. Some people may have a different opinion on it. I would not want to block this change that looks reasonable as is.

**Review by bogdandrutu — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@dashpole I am confused about the changes you mentioned in the description.

**Comment by dashpole** [comment]
@bogdandrutu 
See https://github.com/open-telemetry/opentelemetry-collector/pull/4934.  Bools and Ints are still changed the same way from strings, but for floats, it would change how they are converted to strings in some cases.
